### PR TITLE
[material/menu_anchor.dart] Remove _MenuAnchorState from parent when disposed.

### LINKED ISSUE #149584: [material/menu_anchor.dart] _MenuAnchorState is not always GC'd
### Steps to reproduce

1. Run example.
2. Toggle menu open and closed several times.
3. View _MenuAnchorState instances in Memory devtools.

PR for fix: https://github.com/flutter/flutter/pull/149586

### Expected results

There should only be one `_MenuAnchorState` for each menu anchor

### Actual results

Detached `_MenuAnchorStates` are being retained

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() => runApp(const BugApp());

class BugApp extends StatelessWidget {
  const BugApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(home: Bug());
  }
}

class Bug extends StatefulWidget {
  const Bug({super.key});

  @override
  State<StatefulWidget> createState() => _BugState();
}

class _BugState extends State<Bug> {
  @override
  Widget build(BuildContext context) {
    return Center(
      child: MenuAnchor(
          builder: (context, controller, child) {
            return TextButton(
                onPressed: () {
                  if (controller.isOpen) {
                    controller.close();
                  } else {
                    controller.open();
                  }
                },
                child: const Text('Toggle Menu'));
          },
          menuChildren: [
            SubmenuButton(
              menuChildren: [
                MenuItemButton(
                  onPressed: () {},
                  child: const Text('Item 1'),
                ),
                MenuItemButton(
                  onPressed: () {},
                  child: const Text('Item 2'),
                ),
              ],
              child: const Text('Item 2'),
            ),
            SubmenuButton(
              menuChildren: [
                MenuItemButton(
                  onPressed: () {},
                  child: const Text('Item 1'),
                ),
                MenuItemButton(
                  onPressed: () {},
                  child: const Text('Item 2'),
                ),
              ],
              child: const Text('Item 2'),
            ),
          ]),
    );
  }
}

```

</details>

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/59215665/70853616-058d-4f1f-add4-6ea3de10e264

</details>

### Logs

<details open><summary>Logs</summary>

```console
N/A
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>
I'm running a custom version of Flutter, but this should not affect the bug.

```console

[!] Flutter (Channel master, 3.23.0-13.0.pre.69, on macOS 14.0 23A344 darwin-arm64, locale en-US)
    • Flutter version 3.23.0-13.0.pre.69 on channel master at /Users/davidhicks/development/custom
    ! Warning: `flutter` on your path resolves to /Users/davidhicks/Development/custom/bin/flutter, which
      is not inside your current Flutter SDK checkout at /Users/davidhicks/development/custom. Consider
      adding /Users/davidhicks/development/custom/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/davidhicks/Development/custom/bin/dart, which is
      not inside your current Flutter SDK checkout at /Users/davidhicks/development/custom. Consider
      adding /Users/davidhicks/development/custom/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 124aacbaef (2 days ago), 2024-05-31 22:32:25 -0400
    • Engine revision d81edf635a
    • Dart version 3.5.0 (build 3.5.0-214.0.dev)
    • DevTools version 2.36.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use
      "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/davidhicks/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/davidhicks/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 15.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 15F31d
    • CocoaPods version 1.15.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2023.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.10+0-17.0.10b1087.21-11572160)

[✓] VS Code (version 1.89.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.90.0

[✓] Connected device (3 available)
    • macOS (desktop)                 • macos                 • darwin-arm64   • macOS 14.0 23A344
      darwin-arm64
    • Mac Designed for iPad (desktop) • mac-designed-for-ipad • darwin         • macOS 14.0 23A344
      darwin-arm64
    • Chrome (web)                    • chrome                • web-javascript • Google Chrome
      125.0.6422.113
    ! Error: Browsing on the local area network for David’s iPhone. Ensure the device is unlocked and
      attached with a cable or associated with the same local area network as this Mac.
      The device must be opted into Developer Mode to connect wirelessly. (code -27)

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
```

</details>

### PR DESCRIPTION
**Pull request by davidhicks980**
`_MenuAnchorState` now removes itself from its parent regardless of whether `_MenuAnchorState._isOpen` is true.

Before, _MenuAnchorState would only detach itself when `_MenuAnchorState._isOpen == true`. This led to _MenuAnchorState being retained until an anchor's parent was disposed.

Before:

https://github.com/flutter/flutter/assets/59215665/48f1617a-a3d0-49ab-a767-f41e3a30ea41

After:

https://github.com/flutter/flutter/assets/59215665/31312bb6-49ec-4083-b8de-e9ae2a42a8b3

Fixes https://github.com/flutter/flutter/issues/149584

I was a bit unsure what the best way of testing this change would be, and wanted feedback from the Flutter team. Is there a way for us to view at the instance count of `_MenuAnchorState`, or expose `_MenuAnchorState._anchorChildren.length` for testing?

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by Piinks** [comment]
@davidhicks980 thanks for sending the PR! Would you mind adding a test?

**Comment by polina-c** [comment]
> @polina-c I am surprised the leak tracker did not already catch this. Do you know how we might be able to capture it in a test?

LeakTracker is scoped to catch not disposed leaks at the moment. It does not catch not GCed ones.
It is separate big effort to clean up Flutter Framework from notGCed leaks. And it is not part of leak tracker MVP.
This PR fixes notGCed leak.

Does it help?

**Comment by Piinks** [comment]
Ah ok, thank you for calrifying.

**Comment by Piinks** [comment]
I was looking at the original issue,

> #### Expected results
>There should only be one _MenuAnchorState for each menu anchor
>
> #### Actual results
>Detached _MenuAnchorStates are being retained

Maybe to test @davidhicks980 we could just access `currentState` without needing to expose _MenuAnchorState and just check it is not the same state after opening/closing/opening the menu?

**Comment by davidhicks980** [comment]
@Piinks So I may be wrong about this, but I think the currentState would still correctly report that it has changed since the state attached to the visible MenuAnchor is changing, but the parent _MenuAnchorState would still be holding on to previous instances of the state as well.

To verify, I wrote this test with the unfixed menu and it did pass, but my test could also be flawed:

```dart
testWidgets('Nested anchor state should dispose after a menu is closed', (WidgetTester tester) async {
      dynamic oldState;
      GlobalKey stateKey = GlobalKey();
      await tester.pumpWidget(MaterialApp(
        home: Material(
          child: Center(
            child: MenuAnchor(
               controller: controller,
                builder: (context, controller, child) {
                  return TextButton(
                      onPressed: () {
                        if (controller.isOpen) {
                          controller.close();
                        } else {
                          controller.open();
                        }
                      },
                      child: const Text('Toggle Menu'));
                },
                menuChildren: [
                  SubmenuButton(
                    key: stateKey,
                    menuChildren: [
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 1'),
                      ),
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 2'),
                      ),
                    ],
                    child: const Text('Item 2'),
                  ),
                  SubmenuButton(
                    menuChildren: [
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 1'),
                      ),
                      MenuItemButton(
                        onPressed: () {},
                        child: const Text('Item 2'),
                      ),
                    ],
                    child: const Text('Item 2'),
                  ),
                ]),
          ),
        ),
      ));
      controller.open();
      await tester.pump();
      oldState = stateKey.currentState;
      controller.close();
      await tester.pump();
      controller.open();
      await tester.pump();
      expect(stateKey.currentState, isNot(oldState));
    });
```

**Comment by davidhicks980** [comment]
Okay, assuming checks pass, and forceGC works on CI, this should be good-to-go! I added that forces garbage collection on a weakly-referenced _MenuAnchorState, and it appears to correctly identify if a child state is retained by its parent.

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 1]
```diff
@@ -3803,6 +3804,51 @@ void main() {
         ..rect(color: overlayColor.withOpacity(0.1)),
     );
   });
+
+  // Fixes https://github.com/flutter/flutter/pull/149586
```
Comment (Piinks) [inline packages/flutter/test/material/menu_anchor_test.dart:3808]:
```suggestion
```

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 2]
```diff
@@ -3803,6 +3804,51 @@ void main() {
...
+  testWidgets('Garbage collector destroys child _MenuAnchorState after parent is closed', (WidgetTester tester) async {
+      await tester.pumpWidget(
```
Comment (Piinks) [inline packages/flutter/test/material/menu_anchor_test.dart:3810]:
```suggestion
      // Regression test for https://github.com/flutter/flutter/pull/149586
      await tester.pumpWidget(
```

File: packages/flutter/lib/src/material/menu_anchor.dart [hunk 3]
```diff
@@ -328,8 +328,11 @@ class _MenuAnchorState extends State<MenuAnchor> {
     assert(_debugMenuInfo('Disposing of $this'));
     if (_isOpen) {
       _close(inDispose: true);
-      _parent?._removeChild(this);
     }
+
+    // Fix for https://github.com/flutter/flutter/issues/149584
```
Comment (Piinks) [inline packages/flutter/lib/src/material/menu_anchor.dart:333]:
```suggestion
```

**Review by Piinks — COMMENTED** [review]
Thanks for adding a test - this LGTM with just a few nits

**Review by Piinks — APPROVED** [review]

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 4]
```diff
@@ -3803,6 +3804,51 @@ void main() {
         ..rect(color: overlayColor.withOpacity(0.1)),
     );
   });
+
+  testWidgets('Garbage collector destroys child _MenuAnchorState after parent is closed', (WidgetTester tester) async {
+      // Regression test for https://github.com/flutter/flutter/pull/149586
```
Comment (justinmc) [inline packages/flutter/test/material/menu_anchor_test.dart:3809]: Should you link to the issue? https://github.com/flutter/flutter/issues/149584

**Review by justinmc — APPROVED** [review]
Thank you for fixing this! Good catch. LGTM :+1: 

FYI @gspencergoog because MenuAnchor.

File: packages/flutter/test/material/menu_anchor_test.dart (hunk 4 above)
Comment (Piinks) [inline packages/flutter/test/material/menu_anchor_test.dart:3809]:
```suggestion
      // Regression test for https://github.com/flutter/flutter/issues/149584
```

**Pull request merged**